### PR TITLE
Do not omit predicates when a producer may not have a large enough buffer for a consumer with a non-exact parallel type

### LINKED ISSUE #495: Validation error for non-divisible batch split in layer norm backward
To reproduce, use first commit `52b4422c47c13bf73a90c30d2d321dab62737203` at branch `llu/nondivisible_batch`. The threads_per_block is intentionally set to larger than needed.

validation error if run with:
`./nvfuser_tests --gtest_filter=NVFuserTest.CombinedSchedulerLayerNormBackward_CUDA 2>&1 |tee 1.log`
no error if run with:
`PYTORCH_NVFUSER_DISABLE=predicate_elimination ./nvfuser_tests --gtest_filter=NVFuserTest.CombinedSchedulerLayerNormBackward_CUDA 2>&1 |tee 1.log`

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #495 

Previously, we always insert a predicate for a non-exact parallel type even when the main predicate is determined to be safe to omit. That's been the case since this [PR](https://github.com/csarofeen/pytorch/pull/1580). I believe an only example test case that requires the predicate is `FusionWarpReducePredication`, where `tv3` is not parallelized but its consumer, `tv4`, is parallelized with `TIDx`, which is not unique. `blockDim.x` is in this case 1024, but the corresponding domain of `tv3` is just 50, so we need to predicate the expression for `tv4`. 

This extra predication for non-unique parallel types resulted in incorrect code generation for `FusionMagicSchedulerLayerNormalization_CUDA` as reported in issue #495. That is because the predicate generation for non-unique parallel types [assumes the root domains are predicated](https://github.com/NVIDIA/Fuser/blob/main/csrc/predicate_compute.cpp#L193-L199), but that's not the case when the main predicate is eliminated. This resulted in inconsistent predicates as exhibited in the test, where `T17` is predicated due to `TIDx` being non-unique, whereas its consumer, `T11`, is not since the corresponding domain that is parallelized by `TIDx` is now a root domain of `T11`.

```
T17_l[ iblockIdx.x79{( i1 * ( i3 * i4 ) )}, ithreadIdx.x83{( ceilDiv(( ceilDiv(i5, 1) ), 1) )}rf_p, rS80{1}rf, rUS82{1}rf ] ca_pos( 2 ) produce_pos( 1 )
 root domain : (iS74{i1}, iS75{i3}, iS76{i4}, rS77{i5}rf)
  Outer split: rS77{i5}rf by factor 1 -> rS80{1}rf, iS81{( ceilDiv(i5, 1) )}rf, start offset: 0, stop offset: 0
  Outer split: iS81{( ceilDiv(i5, 1) )}rf by factor 1 -> rUS82{1}rf, ithreadIdx.x83{( ceilDiv(( ceilDiv(i5, 1) ), 1) )}rf_p, start offset: 0, stop offset: 0
 rfactor domain : (iS74{i1}, iS75{i3}, iS76{i4}, rS80{1}rf, rUS82{1}rf, ithreadIdx.x83{( ceilDiv(( ceilDiv(i5, 1) ), 1) )}rf_p)
 contiguity: t t t n n t
  Merge: iS75{i3} and iS76{i4} -> iS78{( i3 * i4 )}
  Merge: iS74{i1} and iS78{( i3 * i4 )} -> iblockIdx.x79{( i1 * ( i3 * i4 ) )}
 leaf domain : (iblockIdx.x79{( i1 * ( i3 * i4 ) )}, ithreadIdx.x83{( ceilDiv(( ceilDiv(i5, 1) ), 1) )}rf_p, rS80{1}rf, rUS82{1}rf)
T11_l[ iblockIdx.x89{( i1 * ( i3 * i4 ) )}, rthreadIdx.x87{( ceilDiv(( ceilDiv(i5, 1) ), 1) )}_p ] ca_pos( 1 ) produce_pos( 2 )
 root domain : (iS84{i1}, iS85{i3}, iS86{i4}, rthreadIdx.x87{( ceilDiv(( ceilDiv(i5, 1) ), 1) )}_p)
 contiguity: t t t n
  Merge: iS85{i3} and iS86{i4} -> iS88{( i3 * i4 )}
  Merge: iS84{i1} and iS88{( i3 * i4 )} -> iblockIdx.x89{( i1 * ( i3 * i4 ) )}
 leaf domain : (iblockIdx.x89{( i1 * ( i3 * i4 ) )}, rthreadIdx.x87{( ceilDiv(( ceilDiv(i5, 1) ), 1) )}_p)
```

It seems there are a couple of fixes we can take. First of all, this assumption that root domains are predicated seems to be fragile as we could potentially change which domains to predicate and this part of the code would need to be  changed as well. Redundant predicates, This was supposed to be an optimization, but may not be relevant anymore as I believe they are now eliminated by index hoisting.

I'll look into the above cleanup as a separate PR, but in this PR, I first changed the predicate elimination logic to also consider if a producer-consumer pair has a pattern like the above. If a producer is not parallelized, thus allocated with the exact size, but a consumer is parallelized by a non-unique parallel type, the predicate is not eliminated. This means that we no longer need to separately add the predicate for non-unique parallel types.

This PR adds a new condition to satisfy in order to eliminate predicates, so it potentially results in less predicates eliminated, but the added condition is quite uncommon, so in reality I expect no significant change. I'll check if any benchmark would result in different code due to this change.

### CODE REVIEW

**Comment by naoyam** [comment]
https://github.com/NVIDIA/Fuser/issues/495#issuecomment-1593441217

`FusionMagicSchedulerLayerNormalization_CUDA` after this PR:

```
  float T14[1];
  T14[0] = 0.00000000000000000e+00;
  if (b2) {
    T14[0]
       = T0[((((T0.stride[3] * ((nvfuser_index_t)threadIdx.x)) + (T0.stride[0] * (((nvfuser_index_t)blockIdx.x) / i0))) + (T0.stride[1] * (i1 / T0.size[2]))) + (T0.stride[2] * (i1 % T0.size[2])))];
  }
  float T11[1];
  T11[0] = 0.00000000000000000e+00;
  float T17[1];
  T17[0] = 0.00000000000000000e+00;
  T17[0]
    = T17[0]
    + T14[0];
  warp::warpReduceTIDX<false, true>(T11[0], T17[0], [](float &a, float b) { a = a + b; }, static_cast<float*>(shared_mem), true, static_cast<float>(0.00000000000000000e+00));
```

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
> I'll check if any benchmark would result in different code due to this change.

Confirmed no benchmark hits the new condition. The generated code does change though since the predicate for non-unique parallel dimensions is not inserted, which is not necessary.

**Review by zasdfgbnm — APPROVED** [review]

**Comment by naoyam** [comment]
!build

File: test/utils.h [hunk 1]
```diff
@@ -137,28 +137,36 @@ inline TensorView* loweredTv(TensorView* tv, GpuLower& gpulw) {
...
+  PredicatedChecker(StmtNameType tv_name, std::vector<Expr*> exprs)
+      : tv_name_(tv_name) {
     kir::IrVisitor::handle(exprs);
   }
 
   using kir::IrVisitor::handle;
   bool is_predicated_ = false;
   bool predicated_ite_ = false;
-  TensorView* tv_ = nullptr;
+  StmtNameType tv_name_ = 0;
```
Comment (zasdfgbnm) [inline test/utils.h:169]: Would fusion managed data make more sense?

Comment (naoyam) [reply test/utils.h:169]: I don't see any particular reason as this is just a quick utility for test validations.

Comment (zasdfgbnm) [reply test/utils.h:169]: When copying a fusion, do we copy the name of IR nodes, or do we reallocate names, and these names coincide?

Comment (zasdfgbnm) [reply test/utils.h:169]: Oh, great. I thought they were just coincide. That's why I felt worried about this approach.

**Pull request merged**